src/mctpd.c

```diff
@@ -2954,6 +3000,16 @@ static int bus_link_set_prop(sd_bus *bus,
 	}
 	lmUserData->role = role.role;
 
+	// Announce on the bus we are endpoint, print warning and ignore error if failed
+	if (lmUserData->role == ENDPOINT_ROLE_ENDPOINT) {
+		rc = notify_discovery(ctx, ifindex);
```

Comment:
This is done unconditionally, but Discovery Notify messages should only be sent on transport types that require them. We may need a way to query the transport type?

Reply:
@jk-ozlabs It would be great if we have that now (we need them to get control message timing information for the recovery mechanism on each interface anyway).

I would propose adding such information to when a binding driver register itself to the MCTP networking subsystem inside the kernel (`mctp_netdev_ops`), like so:

```c
static const struct mctp_netdev_ops mctp_i2c_mctp_ops = {
	.transport = MCTP_TRANSPORT_SMBUS,
	.release_flow = mctp_i2c_release_flow,
};
```

And then exposing them to userspace through netlink.

I don't know how long it would take to merge the patch to linux, but in the meanwhile, we may temporarily provide such information in `mctpd` by basing on the interface name i think. Everything is already somewhat following the convention: `mctpi2c%d`, `mctpserial%d`, ...

I could probably open a PR for this if you think this is ok.

Reply:
Yeah, we'd need the kernel to provide the physical medium identifier. It doesn't belong in the ops struct, but there are certainly ways we can provide that to the `struct mctp_dev` in the kernel (and then on to the netlink interface as a new attribute; say `IFLA_MCTP_PHYS_MEDIUM`).

Given that:

1. the PCIe VDM transport is the only one that requires a Discovery Notify
2. that transport driver is not yet upstream
3. none of the currently-upstream transport drivers provide bindings that require Discovery Notify messages

\- then I would suggest that we add the infrastructure for `IFLA_MCTP_PHYS_MEDIUM`, and have your driver provide the attribute. Then, `mctpd` will only send Discovery Notify messages attribute is present (ie., we know the medium type) and matches the media types that need a Discovery Notify.

Since the Discovery Notify is only required for the non-upstream transport, you'll need kernel patches *anyway*.

> we may temporarily provide such information in mctpd by basing on the interface name i think. Everything is already somewhat following the convention: mctpi2c%d, mctpserial%d, ...

Interface names are not a stable API that we can rely on; they can be renamed arbitrarily.

I'm happy to do the IFLA_MCTP_PHYS_MEDIUM support; drivers will just need a minor change to populate the type value.

Reply:
> I'm happy to do the IFLA_MCTP_PHYS_MEDIUM support; drivers will just need a minor change to populate the type value.

That would be great 😀 Thank you!